src/os/bluestore/BlueStore.cc

```diff
@@ -6932,167 +7070,1527 @@ int BlueStore::expand_devices(ostream& out)
+    ++idx2;
+    size2 += length;
   }
+  real_fm->enumerate_reset();
 
-  int r = _open_db_and_around(false);
-  if (r < 0) {
-    return r;
+  dout(1) << __func__ << "::NCB::size2=" << size2 << ", num2=" << idx2 << dendl;
+  return allocator2;
+}
+
+//---------------------------------------------------------
+// close the active fm and open it in a new mode like makefs()
+// but make sure to mark the full device space as allocated
+// later we will mark all exetents from the allocator as free
+int BlueStore::reset_fm_for_restore()
+{
+  dout(1) << __func__ << "::NCB <<==>> fm->clear_null_manager()" << dendl;
+  fm->shutdown();
+  delete fm;
+  fm = nullptr;
+  freelist_type = "bitmap";
+  KeyValueDB::Transaction t = db->get_transaction();
+  // call _open_fm() with fm_restore set to TRUE
+  // this will mark the full device space as allocated (and not just the reserved space)
+  _open_fm(t, true, true);
+  if (fm == nullptr) {
+    derr << __func__ << "::NCB Failed _open_fm()" << dendl;
+    return -1;
   }
+  db->submit_transaction_sync(t);
+  ceph_assert(!fm->is_null_manager()); 
+  dout(1) << __func__ << "::NCB fm was reactivated in full mode" << dendl;
+  return 0;
+}
 
-  r = _upgrade_super();
-  if (r < 0) {
-    goto out_db;
+
+//---------------------------------------------------------
+// create a temp allocator filled with allocation state from the fm
+// and compare it to the base allocator passed in
+int BlueStore::verify_rocksdb_allocations(Allocator *allocator)
+{
+  dout(1) << __func__ << "::NCB::verify that shared_alloc content is identical to FM" << dendl;
+  // initialize from freelist
+  Allocator* temp_allocator = initialize_allocator_from_freelist(fm);
+  if (temp_allocator == nullptr) {
+    return -1;
   }
 
-  r = _open_collections();
-  if (r < 0)
-    goto out_db;
+  uint64_t insert_count = 0;
+  auto count_entries = [&](uint64_t extent_offset, uint64_t extent_length) {
+    insert_count++;
+  };
+  temp_allocator->dump(count_entries);
+  uint64_t memory_target = cct->_conf.get_val<Option::size_t>("osd_memory_target");
+  int ret = compare_allocators(allocator, temp_allocator, insert_count, memory_target);
 
-  r = _reload_logger();
-  if (r < 0)
-    goto out_coll;
+  delete temp_allocator;
 
-  _kv_start();
+  if (ret == 0) {
+    dout(1) << __func__ << "::NCB::SUCCESS!!! compare(allocator, temp_allocator)" << dendl;
+    return 0;
+  } else {
+    derr << __func__ << "::NCB::**** FAILURE compare(allocator, temp_allocator)::ret=" << ret << dendl;
+    return -1;
+  }
+}
 
-#ifdef HAVE_LIBZBD
-  if (bdev->is_smr()) {
-    _zoned_cleaner_start();
+//---------------------------------------------------------
+// convert back the system from null-allocator to using rocksdb to store allocation
+int BlueStore::push_allocation_to_rocksdb()
+{
+#if 1
+  if (cct->_conf->bluestore_allocation_from_file) {
+    derr << __func__ << "::NCB cct->_conf->bluestore_allocation_from_file must be cleared first" << dendl;
```

Comment:
`bluestore-tool` does not read ceph.conf.
It is impossible to execute `push_allocation_to_rocksdb`.